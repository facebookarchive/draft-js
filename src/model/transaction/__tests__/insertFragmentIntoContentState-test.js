/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 * @oncall draft_js
 */

'use strict';
import type {BlockMap} from 'BlockMap';
import type ContentState from 'ContentState';

const BlockMapBuilder = require('BlockMapBuilder');
const ContentBlock = require('ContentBlock');
const ContentBlockNode = require('ContentBlockNode');
const SelectionState = require('SelectionState');

const getSampleStateForTesting = require('getSampleStateForTesting');
const Immutable = require('immutable');
const insertFragmentIntoContentState = require('insertFragmentIntoContentState');
const invariant = require('invariant');

jest.mock('generateRandomKey');

const {contentState, selectionState} = getSampleStateForTesting();
const {List, Map} = Immutable;

const DEFAULT_BLOCK_CONFIG = {
  key: 'j',
  type: 'unstyled',
  text: 'xx',
  data: Map({a: 1}),
};

const initialBlock = contentState.getBlockMap().first();

const getInvariantViolation = (msg: string) => {
  try {
    /* eslint-disable-next-line */
    invariant(false, msg);
  } catch (e) {
    return e;
  }
};

const createFragment = (
  /* $FlowFixMe[missing-local-annot] The type annotation(s) required by Flow's
   * LTI update could not be added via codemod */
  fragment = {},
  experimentalTreeDataSupport: boolean = false,
) => {
  const ContentBlockNodeRecord = experimentalTreeDataSupport
    ? ContentBlockNode
    : ContentBlock;
  const newFragment = Array.isArray(fragment) ? fragment : [fragment];

  return BlockMapBuilder.createFromArray(
    newFragment.map(
      config =>
        new ContentBlockNodeRecord({
          ...DEFAULT_BLOCK_CONFIG,
          ...config,
        }),
    ),
  );
};

/* $FlowFixMe[missing-local-annot] The type annotation(s) required by Flow's
 * LTI update could not be added via codemod */
const createContentBlockNodeFragment = fragment => {
  return createFragment(fragment, true);
};

const assertInsertFragmentIntoContentState = (
  fragment: BlockMap,
  selection: $FlowFixMe | SelectionState = selectionState,
  content: ContentState = contentState,
) => {
  expect(
    insertFragmentIntoContentState(content, selection, fragment)
      .getBlockMap()
      .toIndexedSeq()
      .toJS(),
  ).toMatchSnapshot();
};

test('must throw if no fragment is provided', () => {
  const fragment = BlockMapBuilder.createFromArray([]);
  expect(() => {
    insertFragmentIntoContentState(contentState, selectionState, fragment);
  }).toThrow();
});

test('must apply fragment to the start', () => {
  assertInsertFragmentIntoContentState(createFragment());
});

test('must apply fragment to within block', () => {
  assertInsertFragmentIntoContentState(
    createFragment(),
    selectionState.merge({
      focusOffset: 2,
      anchorOffset: 2,
      isBackward: false,
    }),
  );
});

test('must apply fragment at the end', () => {
  assertInsertFragmentIntoContentState(
    createFragment(),
    selectionState.merge({
      focusOffset: initialBlock.getLength(),
      anchorOffset: initialBlock.getLength(),
      isBackward: false,
    }),
  );
});

test('must apply multiblock fragments', () => {
  assertInsertFragmentIntoContentState(
    createFragment([
      DEFAULT_BLOCK_CONFIG,
      {
        key: 'k',
        text: 'yy',
        data: Map({b: 2}),
      },
    ]),
  );
});

test('must be able to insert a fragment with a single ContentBlockNode', () => {
  const initialSelection = SelectionState.createEmpty('A');
  const initialContent = contentState.setBlockMap(
    createContentBlockNodeFragment([
      {
        key: 'A',
        text: '',
      },
    ]),
  );

  assertInsertFragmentIntoContentState(
    createContentBlockNodeFragment([
      {
        key: 'B',
        text: 'some text',
      },
    ]),
    initialSelection,
    initialContent,
  );
});

test('must be able to insert fragment of ContentBlockNodes', () => {
  const initialSelection = SelectionState.createEmpty('first');
  const initialContent = contentState.setBlockMap(
    createContentBlockNodeFragment([
      {
        key: 'first',
        text: '',
        nextSibling: 'second',
      },
      {
        key: 'second',
        text: '',
        prevSibling: 'first',
      },
    ]),
  );

  assertInsertFragmentIntoContentState(
    createContentBlockNodeFragment([
      {
        key: 'B',
        text: '',
        children: List(['C']),
        nextSibling: 'E',
      },
      {
        key: 'C',
        parent: 'B',
        text: '',
        children: List(['D']),
      },
      {
        key: 'D',
        parent: 'C',
        text: 'Delta',
      },
      {
        key: 'E',
        text: 'Elephant',
        prevSibling: 'B',
      },
    ]),
    initialSelection,
    initialContent,
  );
});

test('must be able to insert fragment of ContentBlockNodes after nested block', () => {
  const initialSelection = SelectionState.createEmpty('firstChild');
  const initialContent = contentState.setBlockMap(
    createContentBlockNodeFragment([
      {
        key: 'root',
        text: '',
        children: List(['firstChild', 'lastChild']),
      },
      {
        key: 'firstChild',
        parent: 'root',
        text: '',
        nextSibling: 'lastChild',
      },
      {
        key: 'lastChild',
        parent: 'root',
        text: '',
        prevSibling: 'firstChild',
      },
    ]),
  );

  assertInsertFragmentIntoContentState(
    createContentBlockNodeFragment([
      {
        key: 'B',
        text: '',
        children: List(['C']),
        nextSibling: 'E',
      },
      {
        key: 'C',
        parent: 'B',
        text: '',
        children: List(['D']),
      },
      {
        key: 'D',
        parent: 'C',
        text: 'Delta',
      },
      {
        key: 'E',
        text: 'Elephant',
        prevSibling: 'B',
      },
    ]),
    initialSelection,
    initialContent,
  );
});

test('must be able to insert a fragment of ContentBlockNodes while updating the target block with the first fragment block properties', () => {
  const initialSelection = SelectionState.createEmpty('first');
  const initialContent = contentState.setBlockMap(
    createContentBlockNodeFragment([
      {
        key: 'first',
        text: '',
        nextSibling: 'second',
      },
      {
        key: 'second',
        text: '',
        prevSibling: 'first',
      },
    ]),
  );

  assertInsertFragmentIntoContentState(
    createContentBlockNodeFragment([
      {
        key: 'A',
        text: 'Alpha',
        nextSibling: 'B',
      },
      {
        key: 'B',
        text: '',
        children: List(['C']),
        prevSibling: 'A',
      },
      {
        key: 'C',
        parent: 'B',
        text: '',
        children: List(['D']),
      },
      {
        key: 'D',
        parent: 'C',
        text: 'Delta',
      },
    ]),
    initialSelection,
    initialContent,
  );
});

test('must be able to insert a fragment of ContentBlockNodes while updating the target block with the first fragment block properties after nested block', () => {
  const initialSelection = SelectionState.createEmpty('firstChild');
  const initialContent = contentState.setBlockMap(
    createContentBlockNodeFragment([
      {
        key: 'root',
        text: '',
        children: List(['firstChild', 'lastChild']),
      },
      {
        key: 'firstChild',
        parent: 'root',
        text: '',
        nextSibling: 'lastChild',
      },
      {
        key: 'lastChild',
        parent: 'root',
        text: '',
        prevSibling: 'firstChild',
      },
    ]),
  );

  assertInsertFragmentIntoContentState(
    createContentBlockNodeFragment([
      {
        key: 'A',
        text: 'Alpha',
        nextSibling: 'B',
      },
      {
        key: 'B',
        text: '',
        children: List(['C']),
        prevSibling: 'A',
        nextSibling: 'E',
      },
      {
        key: 'C',
        parent: 'B',
        text: '',
        children: List(['D']),
      },
      {
        key: 'D',
        parent: 'C',
        text: 'Delta',
      },
      {
        key: 'E',
        text: 'Elephant',
        prevSibling: 'B',
      },
    ]),
    initialSelection,
    initialContent,
  );
});

test('must throw an error when trying to apply ContentBlockNode fragments when selection is on a block that has children', () => {
  const initialSelection = SelectionState.createEmpty('A');
  const initialContent = contentState.setBlockMap(
    createContentBlockNodeFragment([
      {
        key: 'A',
        text: '',
        children: List(['B']),
      },
      {
        key: 'B',
        text: 'child',
        parent: 'A',
      },
    ]),
  );

  expect(() =>
    insertFragmentIntoContentState(
      initialContent,
      initialSelection,
      createContentBlockNodeFragment([
        {
          key: 'C',
          text: 'some text',
        },
      ]),
    ),
  ).toThrow(
    getInvariantViolation(
      '`insertFragment` should not be called when a container node is selected.',
    ),
  );
});
