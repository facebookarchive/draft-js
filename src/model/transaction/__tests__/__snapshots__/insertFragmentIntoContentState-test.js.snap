// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must apply fragment at the end 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "a",
    "text": "Alphaxx",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
]
`;

exports[`must apply fragment to the start 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "a",
    "text": "xxAlpha",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
]
`;

exports[`must apply fragment to within block 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "a",
    "text": "Alxxpha",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
]
`;

exports[`must apply multiblock fragments 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "a",
    "text": "xx",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {
      "b": 2,
    },
    "depth": 0,
    "key": "key4",
    "text": "yyAlpha",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
]
`;

exports[`must be able to insert a fragment of ContentBlockNodes while updating the target block with the first fragment block properties 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "first",
    "nextSibling": "key15",
    "parent": null,
    "prevSibling": null,
    "text": "Alpha",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "children": Array [
      "key16",
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key15",
    "nextSibling": "second",
    "parent": null,
    "prevSibling": "first",
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "children": Array [
      "key17",
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key16",
    "nextSibling": null,
    "parent": "key15",
    "prevSibling": null,
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key17",
    "nextSibling": null,
    "parent": "key16",
    "prevSibling": null,
    "text": "Delta",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "second",
    "nextSibling": null,
    "parent": null,
    "prevSibling": "key15",
    "text": "",
    "type": "unstyled",
  },
]
`;

exports[`must be able to insert a fragment of ContentBlockNodes while updating the target block with the first fragment block properties after nested block 1`] = `
Array [
  Object {
    "characterList": Array [],
    "children": Array [
      "firstChild",
      "key19",
      "key22",
      "lastChild",
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "root",
    "nextSibling": null,
    "parent": null,
    "prevSibling": null,
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "firstChild",
    "nextSibling": "key19",
    "parent": "root",
    "prevSibling": null,
    "text": "Alpha",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "children": Array [
      "key20",
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key19",
    "nextSibling": "key22",
    "parent": "root",
    "prevSibling": "firstChild",
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "children": Array [
      "key21",
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key20",
    "nextSibling": null,
    "parent": "key19",
    "prevSibling": null,
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key21",
    "nextSibling": null,
    "parent": "key20",
    "prevSibling": null,
    "text": "Delta",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key22",
    "nextSibling": "lastChild",
    "parent": "root",
    "prevSibling": "key19",
    "text": "Elephant",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "lastChild",
    "nextSibling": null,
    "parent": "root",
    "prevSibling": "key22",
    "text": "",
    "type": "unstyled",
  },
]
`;

exports[`must be able to insert a fragment with a single ContentBlockNode 1`] = `
Array [
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "A",
    "nextSibling": null,
    "parent": null,
    "prevSibling": null,
    "text": "some text",
    "type": "unstyled",
  },
]
`;

exports[`must be able to insert fragment of ContentBlockNodes 1`] = `
Array [
  Object {
    "characterList": Array [],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "first",
    "nextSibling": "key6",
    "parent": null,
    "prevSibling": null,
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "children": Array [
      "key7",
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key6",
    "nextSibling": "key9",
    "parent": null,
    "prevSibling": "first",
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "children": Array [
      "key8",
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key7",
    "nextSibling": null,
    "parent": "key6",
    "prevSibling": null,
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key8",
    "nextSibling": null,
    "parent": "key7",
    "prevSibling": null,
    "text": "Delta",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key9",
    "nextSibling": "second",
    "parent": null,
    "prevSibling": "key6",
    "text": "Elephant",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "second",
    "nextSibling": null,
    "parent": null,
    "prevSibling": "key9",
    "text": "",
    "type": "unstyled",
  },
]
`;

exports[`must be able to insert fragment of ContentBlockNodes after nested block 1`] = `
Array [
  Object {
    "characterList": Array [],
    "children": Array [
      "firstChild",
      "key10",
      "key13",
      "lastChild",
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "root",
    "nextSibling": null,
    "parent": null,
    "prevSibling": null,
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "firstChild",
    "nextSibling": "key10",
    "parent": "root",
    "prevSibling": null,
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "children": Array [
      "key11",
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key10",
    "nextSibling": "key13",
    "parent": "root",
    "prevSibling": "firstChild",
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "children": Array [
      "key12",
    ],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key11",
    "nextSibling": null,
    "parent": "key10",
    "prevSibling": null,
    "text": "",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key12",
    "nextSibling": null,
    "parent": "key11",
    "prevSibling": null,
    "text": "Delta",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "key13",
    "nextSibling": "lastChild",
    "parent": "root",
    "prevSibling": "key10",
    "text": "Elephant",
    "type": "unstyled",
  },
  Object {
    "characterList": Array [],
    "children": Array [],
    "data": Object {
      "a": 1,
    },
    "depth": 0,
    "key": "lastChild",
    "nextSibling": null,
    "parent": "root",
    "prevSibling": "key13",
    "text": "",
    "type": "unstyled",
  },
]
`;
